Implement opt.background_thread

interwq opened the pull request:
Added option "background_thread" to enable background async threads, which
handles purging currently. 

This resolves #149.

interwq commented:
I will be leaving this for a while as this depends on #501. 

I'll work on #701 and #598 first and come back to this later.

interwq commented:
I'll keep working on the fast path improvements for the next few days. Will come back and re-open this next week.

interwq commented:
Reopening this as I'm switching back to this part.

interwq commented:
Still missing stats, tests and documentation. But the design is pretty much complete. I feel we should discuss a bit before going further.

Split into two commits:
1) "First step of implementing background threads." includes the basic support for enabling background threads with fixed interval. 
The main thing worth mentioning here is the epoch used for synchronizing thread creation and lifetime, and support toggling on/off opt.background_thread (the mallctl is rw). A global epoch indicates state (even number means enabled), while each background thread updates its index epoch to the current global epoch at the beginning of its execution. Whenever a background thread detects a mismatch between the index epoch it holds and the global epoch, it terminates itself.
 
2) "Implement dynamic interval for background threads."
First change is we keep arena tick running (to advance epoch) even if background thread is enabled, but we do not attempt actual purging in the decay_tick case.
The main change is, we read the backlog and computes how many dirty pages we expect to purge after the next N decay_intervals. We get the value of N through an optimized binary-search.

Any input? Or should I proceed with the remaining details (adding stats, tests and doc)?

interwq commented:
Still updating. Will ping again when ready for review.

interwq commented:
@jasone : this should be ready for more reviews. Some changes regarding the feedback: 1) I simplified the interval search algorithm to try 1 point at a time, instead of 3; and 2) I ended up not abbreviate "background_thread". Other feedback was all addressed accordingly. Also, I squashed the 2 commits into 1. 

From checking the purging progress and intervals / predictions locally with t-test1, I believe this is working as expected.

Still missing tests, doc and some stats I'd like to add.

interwq commented:
Based on offline discussion, decided to use cond var in order to synchronously shutdown background threads.

Still testing.

interwq commented:
@jasone: this should be ready for another review. With introducing cond var, the entire epoch mechanism went away and the logic was simplified (I think).

I plan to add stats, docs and tests in separate commits on top of this.

interwq commented:
@jasone : I think the freebsd test failures are because of the trylock issue you fixed recently -- in this case the pthread_cond_timedwait reports error saying that the mutex was not locked by the owner thread, which indicates malloc_mutex_lock didn't acquire the mutex -- was that the same behavior as you encountered?

davidtgoldblatt commented:
I'll try to do a real review tomorrow, but one thing I noticed looking for the test failure was that this isn't safe w.r.t. spurious wakeups, which pthreads allows.

IIRC, the freebsd bug @jasone found was for a later freebsd version than our CI machines, so I don't think it should cause these test failures.

interwq commented:
re: spurious wakeups, it could cause the threads to run more often than intended, but won't cause failures like this, i.e. return EPERM from pthread_cond_wait, which should only happen when the mutex isn't held by the calling thread.

Actually the freebsd on the worker is pretty recent:
FreeBSD ci-freebsd-a 11.0-RELEASE-p2 FreeBSD 11.0-RELEASE-p2 #0: Mon Oct 24 06:55:27 UTC 2016
So I'm wondering if this is from the same root cause.

davidtgoldblatt commented:
Oh yeah, that's not the cause of these failures; just something I saw when looking for it.

interwq commented:
The test failure earlier was caused by this diff breaking lazy_lock. Just fixed.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,453 @@
+
+/******************************************************************************/
+
+#if !defined(JEMALLOC_BACKGROUND_THREAD)
```
Style nit for simple cpp logic:

  ```#ifndef JEMALLOC_BACKGROUND_THREAD```

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,453 @@
+		return true;
+	}
+
+	for (unsigned i = 0 ; i < ncpus; i++) {
```
```s/0 ;/0;/```

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,453 @@
+#define BILLION UINT64_C(1000000000)
+/* Threshold for determining when to wake up the background thread. */
+#define BACKGROUND_THREAD_NPAGES_THRESHOLD UINT64_C(1024)
+/* Background thread max sleep time is decay_time right shift by this. */
```
```s/right shift/right-shifted/```

jasone replied:
What is the rationale for having a max sleep time?  ```BACKGROUND_THREAD_NPAGES_THRESHOLD``` makes sense to me, but I don't see a conceptual basis for a decay_time/4 limit.

interwq replied:
Oh I'm thinking about an edge case, like there was no activity for a long time (backlog w/ all 0) and we decide to sleep (max), and then workload starts coming in, and we won't be able to pick up until a whole decay period. Does that concern seem reasonable?

jasone replied:
Can we instead ```pthread_cond_signal()``` the background thread on threshold conditions?  Or maybe we should just keep this super simple, and just make the sleep interval fixed: the epoch time or 100ms, whichever is greater.  Another semi-simple option would be to signal the background thread if the number of dirty pages transitions from zero to non-zero, and then use a consistent sleep interval until/unless the number of dirty pages reaches zero again.  The more I think about this, the more the complexity of dynamic sleep intervals worries me.

interwq replied:
Well, I don't feel the fixed-interval solution would really work well, because the deallocation activity doesn't really translate to purge activity directly. With the decay-based algorithm, the majority of the activities happens after tens of epochs, while the first few epochs result in a very small portion to be purged. I still think a dynamic solution works better, since we do have knowledge from the backlog.

If we define some threshold conditions, it's a bit hard to say if we should signal the background thread: if we signal eagerly on zero / non-zero dirty page switch, it could cause many signals, with the background thread not being able to purge anything meaningful. Alternatively, we could read the sleep interval and only decide to signal if the next wake up is far away, but still doesn't feel quite right to me.

jasone replied:
If we could reliably put a lower bound on how many epochs to sleep through, I would agree with you.  However I'm troubled by the need for brittle heuristics that work around our inability to do so (i.e. a huge spike in dirty pages can cause us to change our minds about how long purging can be delayed).

There are two broad operating modes I'm thinking about at the moment:
- No dirty pages (application is idle with respect to deallocation activity).  We don't want our background threads to wake under such conditions.
- Dirty pages (application has recently deallocated).  Waking as often as every 100ms seems totally reasonable to me.

Signal volume shouldn't be a significant overhead; consider that even in the worst case, we would signal once per arena each 100ms.

interwq replied:
I agree that with more dealloc activities, waking up every 100ms sounds very reasonable. However, if we do a fixed interval, on the less-active mode I'm still quite concerned about the wakeup frequency, given that multi-process would cause the wake up times to be multiplied by the number of processes.

I see the point that with the implementation right now we don't have a bound, i.e. we could decide to sleep long and then workload comes in. I think we can fix that part with the following design:
1) gets rid of the BACKGROUND_THREAD_MAX_INTERVAL. This makes the background thread could decide to sleep as long as the decay time.
2) in the interval search algorithm, consider only dirty pages in the backlog, i.e. remove the part that tries to "predict" future dirty pages based on average, in background_thread.c L118-L137.
3) This means background thread itself determines sleep interval solely based on backlog recorded dirty pages. To take future dirty pages into account, when we advance epoch we compute "new dirty pages need to be purged" since last background thread run -- this can be done accurately by walking through the last N epochs, or if we want to avoid the loop we can approximate by utilizing h_steps_sum (avg_new_pages_per_epoch * h_steps_sum[N]).

We do pay the complexity to gain the dynamic interval, but I feel the cost is worth it and makes the option more compelling. Being able to reduce the run frequency considerably also means reduce the interference to application, e.g. when p99 latency is of concern. With a fixed 100ms interval, assuming a two-process host we would still wake up >1k times a second even if no much work present.  With the background_thread option we move the purging or other expensive ops to async threads to avoid doing them on the application threads, so the goals include more predictable behavior and latency. I feel making the interval dynamic matches the motivation there.

Does the above design sound reasonable? If so I can try implementing it to evaluate further.

jasone replied:
I've read your most recent comments several times, and they seem to suggest something like the following, but I'm not sure, so let me describe it in my own words and you can hopefully tell me whether it's the same, or at least similar. :wink:

- The background thread sets its wakeup time based purely on how long it can sleep before needing to purge existing dirty pages (sleeping forever is a valid option).
- When an application thread advances the epoch, it computes whether the background thread is going to wake up soon enough to purge the current set of dirty pages.  If not, it sends a wakeup signal to the background thread, which allows the background thread to recompute and set a new wakeup time.

This has the advantage that background threads can go totally idle (no dirty pages --> sleep forever), but it allows application threads to signal the need for wakeup in a way that limits how often signals are actually sent.

interwq replied:
Yup that's what I was thinking. This way we adds a bit extra work on epoch advancing, but I think I can keep it under control. I do feel the benefit is worth the complexity though. If you think the complexity is acceptable, I'll update the diff that way.

jasone replied:
Sounds great to me.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,453 @@
+/* Background thread max sleep time is decay_time right shift by this. */
+#define BACKGROUND_THREAD_MAX_INTERVAL_SHIFT 2
+/* Minimal sleep interval 200 ms. */
+#define BACKGROUND_THREAD_MIN_INTERVAL_NS (BILLION / 5)
```
This doesn't seem like the right thing to do.  The minimum sleep interval should be one epoch (50ms for default configuration), so that the background thread can purge once per epoch.

interwq replied:
I feel we still should keep a minimal interval to avoid overwhelming wakeups when the decay_time is small, e.g. when we provide sub-second decay_time support, 1 epoch could be very small, e.g. w/ 0.5 second decay time, 2.5ms or 400 times/sec. How about we set the min interval to be 100ms?

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,453 @@
+	}
+	malloc_mutex_postfork_child(tsdn, &background_thread_lock);
+
+	malloc_mutex_lock(tsdn, &background_thread_lock);
```
Locking in the postfork handler is non-idiomatic... it seems strange, maybe dangerous?  Can this code be re-ordered to avoid the locking?

interwq replied:
I did try doing the re-init before the ```malloc_mutex_postfork_child(tsdn, &background_thread_lock);```, however the witness is not happy and says we are not owning ```background_thread_lock``` in one of place asserting ownership. I thought we need to unlock / lock again in order to really "own" the lock after fork?

jasone commented on src/jemalloc.c:
```diff
@@ -1433,6 +1451,24 @@ malloc_init_hard(void) {
+		 */
+		bool err;
+		malloc_mutex_lock(tsd_tsdn(tsd), &background_thread_lock);
+		err = background_thread_create(tsd, 0);
```
Nit: The declaration can be combined into this line.

jasone commented on test/integration/extent.c:
```diff
@@ -2,6 +2,18 @@
+	if (ret == ENOENT) {
+		return false;
+	}
+	assert_d_eq(ret, 0, "unexpected mallctl error");
```
Nit: Capitalize first word of error string.

jasone commented on doc/jemalloc.xml.in:
```diff
@@ -737,6 +737,28 @@ mallctl("arena." STRINGIFY(MALLCTL_ARENAS_ALL) ".decay",
+        linkend="stats.background_thread.num_threads"><mallctl>stats.background_thread</mallctl></link>
+        for related stats.  <link
+        linkend="opt.background_thread"><mallctl>opt.background_thread</mallctl></link>
+        can be used to set the default option.  Note that other types of work
```
IMO we should wait to document other uses until they are implemented.

jasone commented on doc/jemalloc.xml.in:
```diff
@@ -2156,6 +2190,42 @@ struct extent_hooks_s {
+          [<option>--enable-stats</option>]
+        </term>
+        <listitem><para> Number of <link linkend="background_thread">background
+        threads</link> running currently.  Return 0 if
```
```s/Return /Returns /```, here and elsewhere.  That said, I don't think we need to document the 0 return case for any of these, since the meaning of 0 return is self-evident.

interwq commented:
@jasone : as we discussed, this is updated with no upper limit on sleep interval, i.e. if backlog has all 0s, we sleep indefinitely. Alone with this change, on epoch advancement, we call check background thread interval and see if it's necessary to signal the thread (by checking the # of pages we would need to purge till the scheduled wakeup).

Other feedback has all been addressed as well.

I did realize one thing while working on this: since now threads can sleep for ever, it does not solve the inactive arena problem (which was one of the issue we plan to solve with this). If threads are sleep forever, while activities went idle w/o advance epoch, we still got issue there -- to solve this, this is what I have in mind:

We specialize background thread 0 (always alive because of a0): we set its max sleep interval to be the decay time, and have it periodically check other arenas epoch movement, and signal other threads if epoch didn't advance for a long time. This way only 1 thread has to keep waking up on idle, and I believe the inactive issue is handled reasonably.

Does this sound right to you? If so, how about I add it as a separate commit on top of everything else? Other commits are ready for review again.

jasone commented:
Having not yet reviewed the latest version, I don't understand why this approach should be vulnerable to inactive arenas ending up with unpurged memory.  In our earlier discussion we agreed that if an application frees memory that generates unused dirty memory, and that dirty memory would potentially go unpurged for too long due to the background thread's scheduled wakeup time being too far in the future, we could signal the thread so it could recalculate and shorten its sleep time.  Unless there were some threshold below which we don't bother signaling, we would never sleep forever with unpurged dirty memory.

interwq commented:
@jasone : this could only happen if epoch stops advancing due to inactivity. As long as we advance epoch, you are right we signal the threads when reaching the threshold. But in a scenario an arena a bunch of pages w/o advancing epoch, and no active calls afterwards, it will cause the issue.

jasone commented:
Ah, I see.  I was suggesting that we do the check that decides whether to signal the background thread whenever we accrue unused dirty pages, rather than only at epoch advances.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,554 @@
+		decay_purge_interval_search(decay, lb, ub, &next_lb, &next_ub);
+		lb = next_lb;
+		ub = next_ub;
+		assert(++n_search <= 5);
```
Can we make this generic, rather than hard-coded, as some function related to ```lg_floor(SMOOTHSTEP_NSTEPS)``` (and ```decay_time```)?

interwq replied:
Done.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,554 @@
+	/* With the current setting, bisect 5 times reaches ~6 intervals. */
+	unsigned n_search = 0;
+	size_t next_lb, next_ub;
+	while (ub > lb + 7) {
```
What is the intent of the magic number 7 here?  If the intent is to terminate the binary search early to reduce computation, I'd argue that the termination condition should be based on how many elements fall within an acceptable time range, which differs according to both ```SMOOTHSTEP_NSTEPS``` and ```decay_time```.

interwq replied:
Just updated to remove the arbitrary number here. Please see new comments for the changes.

interwq commented:
Yeah that would partly solve the problem. The issue is in that case, even if we wake up the thread, it won't really be able to "see" the newly freed pages, because the epoch hasn't passed yet so it's not recorded in the backlog. Also, checking the wakeup time of the background thread requires the thread mutex, I'm a bit concerned with taking the mutex eagerly on deallocation.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,554 @@
+
+/* Check if we need to signal the background thread early. */
+void
+background_thread_interval_check(tsdn_t *tsdn, arena_t *arena,
```
This function appears to operate as I'd expect, but IMO it should be called whenever the number of unused dirty pages accrued since the last epoch advance or signal reaches a new high water mark.

interwq replied:
Changed.

jasone commented:
Can we make the wakeup time atomic, so that no mutex acquisition is necessary?

Re: the current epoch being in progress, for generality we should consider any associated unused dirty pages (i.e. the newly created ones) as being at the beginning of their decay.  As long as we treat them as such, we never have the potential for sleeping forever with unpurged pages.

interwq commented:
re: dirty pages now treated as in current epoch. This sounds good to me.

re: the mutex -- we can make the wakeup time atomic, but may still got the race condition where we read a non-indefinite sleep time and decide not to wakeup, while the background thread was in the progress of deciding to sleep forever. The synchronization feels pretty vulnerable w/ only atomics. (right now it need both decay and thread mutexes).

Do you think we should still check for background threads interval eagerly? My concern is either mutex overhead, or racy logic w/o mutexes.

interwq commented:
I think we can try this:
Add a simpler version of the current interval check to be called eagerly, and this simple function does: 
```  
if (sleep_time == INDEFINITE) { 
    check_interval(...); // this one does the current full logic
}
```
So this handles the inactive issue lock-free, and does locking when necessary. Does this sound reasonable to you @jasone ?

interwq commented:
@jasone: just updated as we discussed. Added a atomic boolean to indicate whether there is a wakeup scheduled (wakeup time is 64-bit, which would require 64-bit atomic if we switch that). This is added in a separate commit "Check for background thread inactivity on extents_dalloc."

Now I believe the inactivity issue is taken care of: either the application thread will see that the thread is sleeping for ever, or the background thread will see the dirty pages updated by the application threads.

jasone commented on include/jemalloc/internal/background_thread_inlines.h:
```diff
@@ -18,4 +18,42 @@ arena_background_thread_info_get(arena_t *arena) {
+background_thread_wakeup_time_set(tsdn_t *tsdn, background_thread_info_t *info,
+    uint64_t wakeup_time) {
+	malloc_mutex_assert_owner(tsdn, &info->mtx);
+	if (wakeup_time == BACKGROUND_THREAD_INDEFINITE_SLEEP) {
```
Easier to read?:
```C
atomic_store_b(&info->indefinite_sleep,
    wakeup_time == BACKGROUND_THREAD_INDEFINITE_SLEEP, ATOMIC_RELEASE);
```

interwq replied:
Yup. Changed.

jasone commented on src/background_thread.c:
```diff
@@ -495,9 +509,22 @@ background_thread_interval_check(tsdn_t *tsdn, arena_t *arena,
+	if (info->npages_to_purge_new > BACKGROUND_THREAD_NPAGES_THRESHOLD) {
+		should_signal = true;
+	} else if (unlikely(background_thread_indefinite_sleep(info))) {
+		if (extents_npages_get(&arena->extents_dirty) > 0 ||
```
Maybe combine this conditional with the enclosing one?  I don't have strong feelings either way.

interwq replied:
I did start with this branch combined, but looked pretty strange and hard to interpret. Plus the background_thread_indefinite_sleep is kind of more special, so I feel the current way is a bit better.

interwq commented:
Thanks for the quick response and all the feedback @jasone!

Yeah I'm still tweaking the search algorithm in ```arena_decay_compute_purge_interval_impl()``` to incorporate the feedback. I'll let you take another look after I update it there.

interwq commented:
@jasone : Just updated the diff again. Main changes --
1) to address the feedback on ```arena_decay_compute_purge_interval_impl()```: this function is redesigned and refactored. Please take a look again at the function (changed in the first commit). Now the search stopping condition is npurge_upper_bound - npurge_lower_bound < NPAGES_THRESHOLD. 
2) fixed arena_destory and arena_reset: now before doing the destructive work, we terminate the corresponding background thread (a single thread), and re-create afterwards. This seems to be the safest way, also avoids some locking trickery.
Now all tests pass even with background_thread default on.

jasone commented on src/background_thread.c:
```diff
@@ -0,0 +1,572 @@
+
+	size_t lb = BACKGROUND_THREAD_MIN_INTERVAL_NS / decay_interval_ns;
+	size_t ub = SMOOTHSTEP_NSTEPS;
+	/* Decay deadline offset is up to 2 decay intervals. */
```
I don't understand this comment, nor the reason for setting ```lb``` to at least 2.  Is this another case of a magic number that only makes sense under default conditions, or perhaps only for limited numbers of dirty pages?

interwq replied:
Oh this is because we set the epoch deadline to be 1 to 2 intervals here:
https://github.com/jemalloc/jemalloc/blob/dev/src/arena.c#L598

If we don't set it to minimal 2 intervals, we may not advance epoch by the next wakeup, in which case there won't be much useful work to do.

How about we change the comment to 
```
/* Minimal 2 intervals to ensure reaching next epoch deadline. */
```

jasone replied:
Ahh, got it.  The comment update looks good to me.

interwq replied:
Cool. I'll make a couple small tweaks and land this soon.
